consensus_test.go

```diff
@@ -208,6 +129,81 @@ func TestTransition_AcceptState_Validator_VerifyFails(t *testing.T) {
 	})
 }
 
+// Test that if build proposal fails, state machine will change state from AcceptState to RoundChangeState.
+func TestTransition_AcceptState_Proposer_FailedBuildProposal(t *testing.T) {
+	buildProposalFunc := func() (*Proposal, error) {
+		return nil, errors.New("failed to build a proposal")
+	}
+
+	validatorIds := []string{"A", "B", "C"}
+	backend := newMockBackend(validatorIds, nil).HookBuildProposalHandler(buildProposalFunc)
+
+	m := newMockPbft(t, validatorIds, "A", backend)
+	m.state.view = ViewMsg(1, 0)
+	m.setState(AcceptState)
+
+	// Prepare messages
+	m.emitMsg(&MessageReq{
+		From: "A",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(1, 0),
+	})
+	m.emitMsg(&MessageReq{
+		From: "B",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(1, 0),
+	})
+	m.emitMsg(&MessageReq{
+		From: "C",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(1, 0),
+	})
+
+	go func() {
+		for {
+			if m.GetState() == RoundChangeState {
+				m.cancelFn()
+				return
+			}
+		}
+	}()
+
+	m.runCycle(m.ctx)
+	assert.True(t, m.IsState(RoundChangeState))
+}
+
+// Run state machine from AcceptState, proposer node.
+// Artificially induce state machine cancellation and check whether state machine is still in AcceptState.
+func TestTransition_AcceptState_Proposer_Cancellation(t *testing.T) {
+	testAcceptState_Cancellation(t, true)
+}
+
+// Run state machine from AcceptState, non-proposer node.
+// Artificially induce state machine cancellation and check whether state machine is still in the AcceptState.
+func TestTransition_AcceptState_NonProposer_Cancellation(t *testing.T) {
+	testAcceptState_Cancellation(t, false)
+}
+
+func testAcceptState_Cancellation(t *testing.T, isProposerNode bool) {
+	m := newMockPbft(t, []string{"A", "B", "C", "D"}, "D")
+	if !isProposerNode {
+		m.Pbft.state.proposer = "A"
+	}
+
+	m.setState(AcceptState)
+	m.setProposal(&Proposal{
+		Data: mockProposal,
+		Time: time.Now().Add(time.Second),
+	})
+
+	go func() {
+		m.cancelFn()
+	}()
+
+	m.runCycle(m.ctx)
+	assert.True(t, m.IsState(AcceptState), "Expected to be in %s, but was in %s", AcceptState, m.getState())
```

Comment:
Test is good, just remove this assertion. If we cancel the context from the outside, the state is not a result that we expect but just a side effect of the state machine not resetting it and using the default one. So, it is not an output of the cancellation.

Reply:
Ok. I'm just going to add assert.NotPanics, because it would be controversial for me to have a test without any assertions.